executor/cook/executor.py

```diff
@@ -0,0 +1,364 @@
+                return
+            logging.info('{} has not yet been created, sleeping {} ms'.format(self.target_file, sleep_time_ms))
+            time.sleep(sleep_time_ms / 1000.0)  # sleep briefly
+
+        target_file_obj = open(self.target_file, 'r')
+        while True:
+            line = target_file_obj.readline()
+            if not line:
+                if self.completed_signal.isSet():
+                    target_file_obj.close()
+                    break
+                time.sleep(sleep_time_ms / 1000.0)  # sleep briefly
+                continue
+            yield line
+
+    def retrieve_progress_states(self):
+        for line in self.tail(50):
+            progress_report = ProgressWatcher.match_progress_update(line)
+            if progress_report is not None:
+                percent, message = progress_report
+                logging.info('Updating progress to {} percent, message: {}'.format(percent, message))
+                self.progress = {'progress-message': message.strip(), 'progress-percent': int(percent)}
+                yield self.progress
+
+
+class ProgressUpdater(object):
+    """
+        This class is responsible for sending progress updates to the scheduler.
+        It throttles the rate at which progress updates are sent.
+    """
+    def __init__(self, poll_interval_ms):
+        self.poll_interval_ms = poll_interval_ms
+        self.last_reported_time = None
+        self.last_progress_data = None
+        self.lock = Lock()
+
+    def has_enough_time_elapsed_since_last_update(self):
+        if self.last_reported_time is None:
+            return True
+        else:
+            current_time = time.time()
+            time_diff_ms = (current_time - self.last_reported_time) * 1000
+            return time_diff_ms >= self.poll_interval_ms
+
+    def send_progress_update(self, driver, task_id, max_message_length, progress_data, force_send=False):
+        with self.lock:
+            if progress_data is not None and self.last_progress_data != progress_data:
+                if force_send or self.has_enough_time_elapsed_since_last_update():
+                    if type(progress_data) is dict:
+                        logging.info('Sending progress message {}'.format(progress_data))
+                        message_dict = dict(progress_data)
+                        message_dict['task-id'] = task_id
+                        progress_message = json.dumps(message_dict)
+                        send_message(driver, progress_message, max_message_length)
+                        self.last_reported_time = time.time()
+                        self.last_progress_data = progress_data
+                    else:
+                        logging.error('Progress data {} is not a dictionary!'.format(progress_data))
+                else:
+                    logging.error('Not sending progress data as enough time has not elapsed since last update')
```

Comment:
This will be very chatty if the process produces progress updates much more frequently than we poll_interval_ms